torch/autograd/grad_mode.py

```diff
@@ -24,7 +24,7 @@ def __call__(self, func: F) -> F:
 
         @functools.wraps(func)
         def decorate_context(*args, **kwargs):
-            with self.__class__():
+            with deepcopy(self):
```

Comment:
Could you detail how deepcopy is different from creating a new instance here?
In particular this wrapper works just fine for `set_grad_enabled()` that has the same API as inference mode

Reply:
deepcopy will also copy the internal state. This is only useful for `inference_mode`, which has the `mode` attribute (and parameter to `__init__`). If we create a new instance from there (with no arguments), then `inference_mode` will use the default mode and not the one from `self`, which results in the described bug.

For `no_grad`, `enable_grad` and `set_grad_enabled` this does not change the current behaviour.

I guess `copy` instead of `deepcopy` could also work if you see any drawback from `deepcopy`.

Reply:
Ho set_grad_enabled is not a subclass of this. So it would have the same issue if it did.

Reply:
But I'm not sure the deepcopy will work fine here as you don't call the init.
Maybe we want to have a method that each subclass defined to clone itself?

Reply:
It does work fine for the other current subclasses (no_grad and enable_grad) because they either don't have an internal state, or if they have, it is set during `__enter__`, which gets called anyway.

If you expect that in the future there will be another subclass wich does important things in the `__init__` and not in the `__enter__`, then adding `clone` would be a good option too (by default, it would return a new instance with `self.__class__()`, and `inference_mode` would override that).

Reply:
A benefit of the `clone` method solution would also be that we could move `set_grad_enabled` to a subclass of `_DecoratorContextManager` so that it can also be used as a decorator. Let me know if that is preferred.

Reply:
Sorry for the delay, I was out of the office.
I think the clone could be a nice thing here indeed. Do you think you can update this PR to add the clone method on the main class?

Reply:
Done. Also squashed and rebased on master